[quant][graphmode][fix] Fold conv bn

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40889 [quant][graphmode][fix] Print the node in error message
* **#40865 [quant][graphmode][fix] Fold conv bn**
* #40743 [quant][graphmode] Support quantization for `aten::apend`

Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22338074](https://our.internmc.facebook.com/intern/diff/D22338074)

jerryzh168 committed:
[quant][graphmode][fix] Fold conv bn

Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Fold conv bn"


Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Fold conv bn"


Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/csrc/jit/passes/fold_conv_bn.cpp:
```diff
@@ -113,8 +113,8 @@ class FoldConvBatchNormHelper {
       conv_module_and_params_;
-  std::unordered_map<Graph*, std::vector<std::tuple<std::string, std::string>>>
-      conv_bn_names_;
+  std::unordered_map<Graph*, std::vector<std::tuple<std::vector<std::string>, std::vector<std::string>>>>
```
this is getting a bit harder to read, perhaps we can define a type for this and/or add a comment describing what this contains?  Otherwise the reader has to read the code line by line.

jerryzh168 replied:
will do

vkuzo reviewed (approved):
lg!

Just for my own understanding, why did the existing `is_conv2d_module, is_batchnorm2d_module` filters not handle this case?

jerryzh168 commented:
> lg!
> 
> Just for my own understanding, why did the existing `is_conv2d_module, is_batchnorm2d_module` filters not handle this case?

theses are not run previously, I added the code to run these filters in this PR.

jerryzh168 committed:
Update on "[quant][graphmode][fix] Fold conv bn"


Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22338074](https://our.internmc.facebook.com/intern/diff/D22338074)

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/fold_conv_bn.cpp:
```diff
@@ -251,29 +251,28 @@ void FoldConvBatchNormHelper::analyze(
-        conv_bn_names_[g] = {};
+        conv_bn_paths_[g] = {};
         for (const Match& match : matches) {
+          if (!std::all_of(
```
Instead of filtering the matches here, can we change the pattern definitions?
```c++
 const PatternInfo pattern2d = PatternInfo::parse_from_str(
      R"(	   
graph(%self, %input, %conv, %batchnorm):	
    %conv_out = prim::CallMethod[name="forward"](%conv, %input)	   
    %bn_out = prim::CallMethod[name="forward"](%batchnorm, %conv_out)	
    return (%bn_out))",	 
      {is_conv2d_module, is_batchnorm2d_module});	   

```
So that the module type is also checked? Currently the pattern seems to check for two forwards with the output of the first forward going as input to the second, which is generic.

jerryzh168 replied:
ClassType definition is not supported in subgraph matcher right now, the filtering will make sure we only match desired types

jerryzh168 committed:
Update on "[quant][graphmode][fix] Fold conv bn"


Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22338074](https://our.internmc.facebook.com/intern/diff/D22338074)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Fold conv bn"


Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22338074](https://our.internmc.facebook.com/intern/diff/D22338074)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Fold conv bn"


Summary:
1. applied filter for the module types
2. removed the assumption that the conv bn are immediate child of parent module

Test Plan:
python test/test_quantization.py TestQuantizeJitPasses
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22338074](https://our.internmc.facebook.com/intern/diff/D22338074)

[ghstack-poisoned]
